### PR TITLE
fix: replace with getGlobal

### PR DESCRIPTION
**Pull request by fw6**
In some runtime environments(the Mini Program unique to China), self or global may be empty.

### CODE REVIEW

**Comment by davidkpiano** [comment]
This seems to be missing the `getGlobal()` implementation?

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis#search_for_the_global_across_environments

**Comment by fw6** [comment]
> This seems to be missing the `getGlobal()` implementation?
> 
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis#search_for_the_global_across_environments

:construction: the Mini Program development framework don't provide `global/self`, only `window` can get [(framework app-service)](https://developers.weixin.qq.com/miniprogram/en/dev/framework/app-service/).  

🚑️  It should be better to replace that line of code with function `getGlobal`.

iShot2022-03-08 21 04 36

**Comment by fw6** [comment]
> This seems to be missing the `getGlobal()` implementation?
> 
> https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis#search_for_the_global_across_environments

:construction: the Mini Program development framework don't provide `global/self`, only `window` can get [(framework app-service)](https://developers.weixin.qq.com/miniprogram/en/dev/framework/app-service/).  

🚑️  It should be better to replace that line of code with function `getGlobal`.

iShot2022-03-08 21 04 36

File: packages/core/src/interpreter.ts [hunk 1]
```diff
@@ -142,9 +142,9 @@ export class Interpreter<
         return clearTimeout(id);
       }
     } as Clock,
-    logger: global.console.log.bind(console),
+    logger: global?.console.log.bind(console),
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:145]: Is this change required though? Are all standard~ global objects not available in this environment? Even `self` and `globalThis`?

Comment (fw6) [reply packages/core/src/interpreter.ts:145]: my fault😢 . this change is not required. 
Only `self` and `globalThis` is undefined. `window` is available.

Comment (fw6) [reply packages/core/src/interpreter.ts]: Restored that line. :fire:

Comment (fw6) [reply packages/core/src/interpreter.ts:145]: `globalThis` may be `undefined`

image
1

Comment (Andarist) [reply packages/core/src/interpreter.ts:145]: > window is available.

This is somewhat confusing as on the screenshot shared [here](https://github.com/statelyai/xstate/pull/3133#issuecomment-1061755747) it states that `window` shouldn't be available in this environment 🤔

Comment (fw6) [reply packages/core/src/interpreter.ts:145]: replacing `global` with `window` or `globalThis` may be the right way, but why not use `getGlobal()` directly.

image

Comment (fw6) [reply packages/core/src/interpreter.ts:145]: `global` isn't defined on this file, so it refers to `NodeJS.Global` here(that's the reason no error in TS)

Comment (Andarist) [reply packages/core/src/interpreter.ts:145]: > replacing global with window or globalThis may be the right way, but why not use getGlobal() directly.

I think you might have misunderstood me - I'm totally on board with using `getGlobal` here. I'm just wondering what global object is gonna be available in your environment.

Comment (fw6) [reply packages/core/src/interpreter.ts:145]: 😅  `global` don't have the `console` object
image

Comment (Andarist) [reply packages/core/src/interpreter.ts:145]: That is... some bizarre env there 😅 The especially surprising bit is this:
```
window "[object Object]"
globalThis "[object Window]"
```

I'm still surprised that `window` is available here as that contradicts the documentation - but, of course, I believe the screenshot that you have shared just now. Is `globalThis.console` available there?

Comment (fw6) [reply packages/core/src/interpreter.ts:145]: >  Is `globalThis.console` available there?

Yep.

Comment (Andarist) [reply packages/core/src/interpreter.ts]: I've tweaked the fix and pushed it out to your branch. Could you check if this variant works for you?

File: packages/core/src/interpreter.ts [hunk 2]
```diff
@@ -142,9 +142,9 @@ export class Interpreter<
         return clearTimeout(id);
       }
     } as Clock,
-    logger: global.console.log.bind(console),
```
Comment (Andarist) [inline packages/core/src/interpreter.ts:145]: my rationale here is that the argument passed to this `bind` call wasn't looked up on the `global` variable anyway. This means that all our users were already accessing this in their respective global scopes automatically and it worked. it had to work as otherwise they wouldn't be able to run XState at all as it would throw `Uncaught ReferenceError: console is not defined` at the initialization time.

So I think that we don't have to resolve the global object here manually at all and we can just rely on the standard JS scope lookup.

File: packages/core/src/interpreter.ts (hunk 1 above)
Comment (fw6) [reply packages/core/src/interpreter.ts]: Everything works perfectly. Thank you so much :tada: :rocket:
